libs/langchain/langchain/schema/runnable/history.py

```diff
@@ -0,0 +1,254 @@
+
+    Example:
+        .. code-block:: python
+
+            from langchain.schema.prompts import ChatPromptTemplate, MessagesPlaceholder
+            from langchain.chat_models import ChatAnthropic
+            from langchain.schema.runnables import RunnableWithMessageHistory
+
+    """
+
+    factory: Callable[[str], BaseChatMessageHistory]
+    input_messages_key: Optional[str] = None
+    output_messages_key: Optional[str] = None
+    message_history_key: Optional[str] = None
+
+    def __init__(
+        self,
+        runnable: Runnable[
+            MessagesOrDictWithMessages,
+            Union[str, BaseMessage, MessagesOrDictWithMessages],
+        ],
+        factory: Callable[[str], BaseChatMessageHistory],
+        *,
+        input_messages_key: Optional[str] = None,
+        output_messages_key: Optional[str] = None,
+        message_history_key: Optional[str] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize RunnableWithMessageHistory.
+
+        Args:
+            runnable: The base Runnable to be wrapped.
+
+                Must take as input one of:
+                - A sequence of BaseMessages
+                - A dict with one key for all messages
+                - A dict with one key for the current input string/message(s) and
+                    a separate key for historical messages. If the input key points
+                    to a string, it will be treated as a HumanMessage in history.
+
+                Must return as output one of:
+                - A string which can be treated as an AIMessage
+                - A BaseMessage or sequence of BaseMessages
+                - A dict with a key for a BaseMessage or sequence of BaseMessages
+
+            factory: Function that returns a new BaseChatMessageHistory given a
+                session id. Should take a single parameter `session_id` which is a
+                string.
+            input_messages_key: Must be specified if the base runnable accepts a dict
+                as input.
+            output_messages_key: Must be specified if the base runnable returns a dict
+                as output.
+            message_history_key: Must be specified if the base runnable accepts a dict
+                as input and expects a separate key for historical messages.
+            **kwargs: Arbitrary additional kwargs to pass to parent class
+                ``RunnableBindingBase`` init.
+        """
+        history_chain: Runnable = RunnableLambda(
+            self._enter_history, self._aenter_history
+        ).with_config(run_name="load_history")
+        messages_key = message_history_key or input_messages_key
+        if messages_key:
+            history_chain = RunnablePassthrough.assign(
+                **{messages_key: history_chain}
+            ).with_config(run_name="insert_history")
+        bound = (
+            history_chain | runnable.with_listeners(on_end=self._exit_history)
```

Comment:
Do we need to support an async version of on_end ?

Reply:
With listeners runs it in thread pool. Given chat history classes don't have async methods unfortunately there's no advantage to having native async method